### PR TITLE
Add response statistics to GRPC frontend

### PR DESCRIPTION
**Pull request by Tabrizian**

### CODE REVIEW

File: protobuf/grpc_service.proto [hunk 1]
```diff
@@ -1104,6 +1159,17 @@ message ModelStatistics
   //@@     compute).
   //@@
   repeated InferBatchStatistics batch_stats = 7;
+
+  //@@  .. cpp:var:: InferResponseStatistics response_stats (repeated)
+  //@@
+  //@@     The aggregate statistics for different responses that the 
+  //@@     model may produce. The response statistics can provide the
+  //@@     breakdown of the time it takes for the backend to generate
```
Comment (tanmayv25) [inline protobuf/grpc_service.proto:1167]: I don't think you can obtain breakdown for each individual response. These stats are accumulated. At most you can obtain the average of the breakdown across the i'th responses for all the requests that generated same number of responses.

File: protobuf/grpc_service.proto [hunk 2]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+//@@.. cpp:var:: message InferResponseStatistics
+//@@
+//@@   Inference response statistics.
+//@@
+message InferResponseStatistics
+{
+  //@@
+  //@@.. cpp:var:: message InferResponseStat (repeated)
+  //@@
+  //@@   Cumalative statistics corresponding to responses generated by the
```
Comment (tanmayv25) [inline protobuf/grpc_service.proto:1085]: cumulative

File: protobuf/grpc_service.proto [hunk 3]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+  //@@   requests that create certain number of responses. For example, each
```
Comment (tanmayv25) [inline protobuf/grpc_service.proto:1086]: Move the example part to L1170 and make the it more general like:

In other words, response statistics for requests that generate identical number of responses will be accumulated in a single InferResponseStatistics message.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]
We would have to update 
https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_statistics.md

We can add some examples for the response statistics in the document.

File: protobuf/grpc_service.proto [hunk 4]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+  //@@
+  //@@     Cumulative count and duration for successful inference
+  //@@     response.
+  //@@
+  StatisticDuration success = 3;
+
+  //@@  .. cpp:var:: StatisticDuration fail
+  //@@
+  //@@     Cumulative count and duration for failed inference
+  //@@     response.
```
Comment (fpetrini15) [inline protobuf/grpc_service.proto:1070]: responses

File: protobuf/grpc_service.proto [hunk 5]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+//@@   Inference response statistics.
+//@@
+message InferResponseStatistics
+{
+  //@@
+  //@@.. cpp:var:: message InferResponseStat (repeated)
+  //@@
+  //@@   Cumalative statistics corresponding to responses generated by the
+  //@@   requests that create certain number of responses. For example, each
+  //@@   InferResponseStatistics contains the cumalative statistics for all the
```
Comment (fpetrini15) [inline protobuf/grpc_service.proto:1087]: cumulative

**Review by fpetrini15 — CHANGES_REQUESTED** [review]

File: protobuf/grpc_service.proto [hunk 6]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+//@@
+//@@.. cpp:var:: message InferResponseStat
+//@@
+//@@   Inference response statistics.
+//@@
+message InferResponseStat 
+{
+  //@@  .. cpp:var:: StatisticDuration compute_infer
+  //@@
+  //@@     The count and cumulative duration generate the inference response.
```
[the remark this answers is not in the document]
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto [hunk 7]
```diff
@@ -1035,6 +1035,61 @@ message InferBatchStatistics
...
+  //@@     produced by the model framework / backend for the given response index.
+  //@@     For example, this duration should include the time to copy output
+  //@@     tensor data from the GPU.
+  //@@
+  StatisticDuration compute_output = 2;
+
+  //@@  .. cpp:var:: StatisticDuration success
+  //@@
+  //@@     Cumulative count and duration for successful inference
+  //@@     response.
```
[the remark this answers is not in the document]
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto (hunk 1 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto (hunk 2 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto (hunk 3 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto (hunk 4 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

File: protobuf/grpc_service.proto (hunk 5 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Updated.

**Comment by kthui** [comment]
Rebased.
Changes since last review:
```
$ git diff dac330fe7583e375e46d25c2d4980ada7491d92f ed7e48dfb4e23ad07c8ec4f062789c158724be95
diff --git a/protobuf/grpc_service.proto b/protobuf/grpc_service.proto
index c945b6a..0ec20da 100644
--- a/protobuf/grpc_service.proto
+++ b/protobuf/grpc_service.proto
@@ -1044,14 +1044,15 @@ message InferResponseStat
 {
   //@@  .. cpp:var:: StatisticDuration compute_infer
   //@@
-  //@@     The count and cumulative duration generate the inference response.
+  //@@     The count and cumulative duration that generate the inference
+  //@@     results for a response.
   //@@
   StatisticDuration compute_infer = 1;
 
   //@@  .. cpp:var:: StatisticDuration compute_output
   //@@
-  //@@     The count and cumulative duration to extract output tensor data
-  //@@     produced by the model framework / backend for the given response index.
+  //@@     The count and cumulative duration to extract the output tensor data
+  //@@     from the inference result for a response.
   //@@     For example, this duration should include the time to copy output
   //@@     tensor data from the GPU.
   //@@
@@ -1060,14 +1061,14 @@ message InferResponseStat
   //@@  .. cpp:var:: StatisticDuration success
   //@@
   //@@     Cumulative count and duration for successful inference
-  //@@     response.
+  //@@     responses.
   //@@
   StatisticDuration success = 3;
 
   //@@  .. cpp:var:: StatisticDuration fail
   //@@
   //@@     Cumulative count and duration for failed inference
-  //@@     response.
+  //@@     responses.
:...skipping...
diff --git a/protobuf/grpc_service.proto b/protobuf/grpc_service.proto
index c945b6a..0ec20da 100644
--- a/protobuf/grpc_service.proto
+++ b/protobuf/grpc_service.proto
@@ -1044,14 +1044,15 @@ message InferResponseStat
 {
   //@@  .. cpp:var:: StatisticDuration compute_infer
   //@@
-  //@@     The count and cumulative duration generate the inference response.
+  //@@     The count and cumulative duration that generate the inference
+  //@@     results for a response.
   //@@
   StatisticDuration compute_infer = 1;
 
   //@@  .. cpp:var:: StatisticDuration compute_output
   //@@
-  //@@     The count and cumulative duration to extract output tensor data
-  //@@     produced by the model framework / backend for the given response index.
+  //@@     The count and cumulative duration to extract the output tensor data
+  //@@     from the inference result for a response.
   //@@     For example, this duration should include the time to copy output
   //@@     tensor data from the GPU.
   //@@
@@ -1060,14 +1061,14 @@ message InferResponseStat
   //@@  .. cpp:var:: StatisticDuration success
   //@@
   //@@     Cumulative count and duration for successful inference
-  //@@     response.
+  //@@     responses.
   //@@
   StatisticDuration success = 3;
 
   //@@  .. cpp:var:: StatisticDuration fail
   //@@
   //@@     Cumulative count and duration for failed inference
-  //@@     response.
+  //@@     responses.
   //@@
   StatisticDuration fail = 4;
 }
@@ -1082,10 +1083,8 @@ message InferResponseStatistics
   //@@
   //@@.. cpp:var:: message InferResponseStat (repeated)
   //@@
-  //@@   Cumalative statistics corresponding to responses generated by the
-  //@@   requests that create certain number of responses. For example, each
-  //@@   InferResponseStatistics contains the cumalative statistics for all the
-  //@@   requests that generate 5 responses.
+  //@@   Cumulative statistics corresponding to responses generated by the
+  //@@   requests that create certain number of responses.
   //@@
   repeated InferResponseStat responses = 1;
 }
@@ -1162,12 +1161,13 @@ message ModelStatistics
 
   //@@  .. cpp:var:: InferResponseStatistics response_stats (repeated)
   //@@
-  //@@     The aggregate statistics for different responses that the 
-  //@@     model may produce. The response statistics can provide the
-  //@@     breakdown of the time it takes for the backend to generate
-  //@@     each individual response. The results in the response statistics
-  //@@     are aggregated by the number of responses that each request
-  //@@     produces and the response index.
+  //@@     The aggregate statistics for different responses that the model may
+  //@@     produce. The response statistics can provide the average of the time
+  //@@     it takes for the backend to generate each individual response. The
+  //@@     results in the response statistics are aggregated by the number of
+  //@@     responses that each request produces and the response index.
+  //@@     For example, each InferResponseStatistics contains the cumulative
+  //@@     statistics for all the requests that generate 5 responses.
   //@@
   repeated InferResponseStatistics response_stats = 8;
 }
$ 
```

File: protobuf/grpc_service.proto [hunk 8]
```diff
@@ -1104,6 +1158,25 @@ message ModelStatistics
...
   repeated InferBatchStatistics batch_stats = 7;
+
+  //@@  .. cpp:var:: InferResponseStatistics response_stats (repeated)
+  //@@
+  //@@     The aggregate statistics for different responses that the model may
+  //@@     produce. The response statistics can provide the average of the time
+  //@@     it takes for the backend to generate each individual response. The
+  //@@     results in the response statistics are aggregated by the number of
+  //@@     responses that each request produces and the response index.
+  //@@     For example, each InferResponseStatistics contains the cumulative
```
Comment (tanmayv25) [inline protobuf/grpc_service.proto:1169]: This line is not very accurate. Please generalize it in better way. And provide complete expected response_stats.
For example, take a model that only produces following number of responses: N, M. 
How would the response_stats look for this case?

File: protobuf/grpc_service.proto [hunk 9]
```diff
@@ -1035,6 +1035,60 @@ message InferBatchStatistics
...
+//@@
+//@@   Inference response statistics.
+//@@
+message InferResponseStatistics
+{
+  //@@
+  //@@.. cpp:var:: message InferResponseStat (repeated)
+  //@@
+  //@@   Cumulative statistics corresponding to responses generated by the
+  //@@   requests that create certain number of responses.
```
Comment (rmccorm4) [inline protobuf/grpc_service.proto:1087]: This is a bit confusing to read. Are these cumulative for a single request with multiple responses? Or cumulative across all responses?

File: protobuf/grpc_service.proto [hunk 10]
```diff
@@ -1035,6 +1035,60 @@ message InferBatchStatistics
...
+//@@.. cpp:var:: message InferResponseStat
+//@@
+//@@   Inference response statistics.
+//@@
+message InferResponseStat 
+{
+  //@@  .. cpp:var:: StatisticDuration compute_infer
+  //@@
+  //@@     The count and cumulative duration that generate the inference
+  //@@     results for a response.
```
Comment (rmccorm4) [inline protobuf/grpc_service.proto:1048]: Is this cumulative or for a single request/response? Can you share example statistics output?

File: protobuf/grpc_service.proto (hunk 8 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Yes. It is rewritten.

File: protobuf/grpc_service.proto (hunk 9 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: Yes. It is rewritten.

Comment (kthui) [reply protobuf/grpc_service.proto]: It is cumulative across multiple responses for a request.

File: protobuf/grpc_service.proto (hunk 10 above)
Comment (kthui) [reply protobuf/grpc_service.proto]: It is cumulative across multiple responses for a request.

Comment (kthui) [reply protobuf/grpc_service.proto]: It is rewritten.

**Comment by kthui** [comment]
The comments are rewritten. Are they clearer than the previous one?

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**